homeassistant/components/sabnzbd/__init__.py

```diff
@@ -147,88 +39,50 @@ class SabnzbdSensorEntityDescription(SensorEntityDescription, SabnzbdRequiredKey
-        return False
-
-
-async def async_configure_sabnzbd(
-    hass, config, use_ssl, name=DEFAULT_NAME, api_key=None
-):
-    """Try to configure Sabnzbd and request api key if configuration fails."""
-
-    host = config[CONF_HOST]
-    port = config[CONF_PORT]
-    web_root = config.get(CONF_PATH)
-    uri_scheme = "https" if use_ssl else "http"
-    base_url = BASE_URL_FORMAT.format(uri_scheme, host, port)
-    if api_key is None:
-        conf = await hass.async_add_executor_job(
-            load_json, hass.config.path(CONFIG_FILE)
-        )
-        api_key = conf.get(base_url, {}).get(CONF_API_KEY, "")
-
-    sab_api = SabnzbdApi(
-        base_url, api_key, web_root=web_root, session=async_get_clientsession(hass)
-    )
-    if await async_check_sabnzbd(sab_api):
-        async_setup_sabnzbd(hass, sab_api, config, name)
-    else:
-        async_request_configuration(hass, config, base_url, web_root)
-
-
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the SABnzbd component."""
+    hass.data.setdefault(DOMAIN, {})
 
-    async def sabnzbd_discovered(service: str, info: DiscoveryInfoType | None) -> None:
-        """Handle service discovery."""
-        if not info:
-            return
-        ssl = info.get("properties", {}).get("https", "0") == "1"
-        await async_configure_sabnzbd(hass, info, ssl)
+    if hass.config_entries.async_entries(DOMAIN):
+        return True
 
-    discovery.async_listen(hass, SERVICE_SABNZBD, sabnzbd_discovered)
+    if DOMAIN in config:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=config[DOMAIN],
+            )
+        )
 
-    if (conf := config.get(DOMAIN)) is not None:
-        use_ssl = conf[CONF_SSL]
-        name = conf.get(CONF_NAME)
-        api_key = conf.get(CONF_API_KEY)
-        await async_configure_sabnzbd(hass, conf, use_ssl, name, api_key)
     return True
 
 
-@callback
-def async_setup_sabnzbd(hass, sab_api, config, name):
-    """Set up SABnzbd sensors and services."""
-    sab_api_data = SabnzbdApiData(sab_api, name, config.get(CONF_SENSORS, {}))
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the SabNzbd Component."""
+    sab_api = await get_client(hass, entry.data)
+    if not sab_api:
+        raise ConfigEntryNotReady
 
-    if config.get(CONF_SENSORS):
-        hass.data[DATA_SABNZBD] = sab_api_data
-        hass.async_create_task(
-            discovery.async_load_platform(hass, Platform.SENSOR, DOMAIN, {}, config)
-        )
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
+        KEY_API: sab_api,
+        KEY_NAME: entry.data[CONF_NAME],
+    }
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    sab_api_data = SabnzbdApiData(sab_api)
 
     async def async_service_handler(service: ServiceCall) -> None:
```

Comment:
If we allow more than one config entry, we need to identify which config entry, ie which api instance, is targeted during service calls, and also not register the services if they are already registered.

Preferably also unregister the services when the last loaded config entry is unloaded.

Reply:
Here's an example of how the Simplisafe integration does this. How to identify the api depends on what identifiers are available.

https://github.com/home-assistant/core/blob/21f9b19fad8c61db87e4ece97684a48a287ce4d8/homeassistant/components/simplisafe/__init__.py#L330-L394

https://github.com/home-assistant/core/blob/21f9b19fad8c61db87e4ece97684a48a287ce4d8/homeassistant/components/simplisafe/__init__.py#L426-L435